theano/tensor/opt.py

```diff
@@ -502,9 +504,27 @@ def register(inner_lopt):
         return lopt
 
 
-# Register merge_optimizer as a global opt during canonicalize
-compile.optdb['canonicalize'].register('canon_merge', merge_optimizer,
-                                       'fast_run', final_opt=True)
+# Register in the canonizer Equilibrium as a local opt the merge opt.
+# Without this, as the equilibrium have ignore_newtrees=False, we
+# won't merge all nodes if it is set as a global optimizer with
+# final_opt=True.
+#
+# This work due to those properties:
+# 1) the EQ will execute first the optimizer that trac all nodes.
+# 2) after an local optimization being applied, if the
+#    current node is still in the graph, it will continue to the next
+#    local optimizer. So this won't trigger more iteration.
+def add_merge_feature(fgraph):
+    if not hasattr(fgraph, 'merge_feature'):
+        fgraph.attach_feature(theano.gof.opt.MergeFeature())
+
+
+@register_canonicalize('fast_compile', 'merge')
+@gof.local_optimizer(None, requirements=[add_merge_feature])
```

Comment:
Do we really need to change the signature of `gof.local_optimizer` for that? Could we use `AddFeatureOptimizer` (currently in `theano/compile/mode.py`) instead to make sure the feature is there?